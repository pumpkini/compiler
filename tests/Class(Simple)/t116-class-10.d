
class A {
    Z a(Z z) {
        Print("entering method a of A");
        return z.c();
    }

    Z b(Z z) {
        Print("entering method b of A");
        return z.c();
    }
}

class Z {

    int counter;

    void init() {
        counter = 0;
    }

    Z c() {
        Print("inside method c of Z");
        counter = counter + 1;
        return this;
    }

    void print() {
        Print("Value of counter is: ", counter);
    }
}

int main() {
    A a;
    Z z;
    a = new A;
    z = new Z;
    z.init();

    a.b(a.a(a.b(z)));
    z.print();
}
