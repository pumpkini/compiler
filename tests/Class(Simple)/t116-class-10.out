entering method b of A
inside method c of Z
entering method a of A
inside method c of Z
entering method b of A
inside method c of Z
Value of counter is: 3
